client.go

```diff
@@ -0,0 +1,280 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package beanq
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/retail-ai-inc/beanq/helper/stringx"
+	"github.com/retail-ai-inc/beanq/helper/timex"
+	"github.com/rs/xid"
+)
+
+// subscribe type
+type subscribeType int
+
+const (
+	normalSubscribe subscribeType = iota + 1
+	sequentialSubscribe
+)
+
+// message type
+type moodType string
+
+const (
+	NORMAL     moodType = "normal"
+	DELAY      moodType = "delay"
+	SEQUENTIAL moodType = "sequential"
+)
+
+type (
+	// IBaseCmd BaseCmd public method
+	IBaseCmd interface {
+		filter(message *Message) error
+	}
+	// IBaseSubscribeCmd BaseSubscribeCmd subscribe method
+	IBaseSubscribeCmd interface {
+		IBaseCmd
+		init(broker IBroker) *Subscribe
+		Run(ctx context.Context)
+	}
+	// define available command
+	cmdAble func(ctx context.Context, cmd IBaseCmd) error
+	// Client beanq's client
+	Client struct {
+		message *Message
+		cmdAble
+		broker IBroker
+	}
+)
+
+func New(config *BeanqConfig) *Client {
+	// init config,Will merge default options
+	config.init()
+
+	client := &Client{}
+	client.message = &Message{
+		Id:        "",
+		Topic:     config.Topic,
+		Channel:   config.Channel,
+		MaxLen:    config.MaxLen,
+		Retry:     config.JobMaxRetries,
+		Priority:  config.Priority,
+		TimeToRun: config.TimeToRun,
+	}
+	client.cmdAble = client.process
+	client.broker = NewBroker(config)
+	return client
+}
+
+func (t *Client) SetId(id string) *Client {
+	t.message.Id = id
+	return t
+}
+
+func (t *Client) Channel(name string) *Client {
+	t.message.Channel = name
+	return t
+}
+
+func (t *Client) Topic(name string) *Client {
+	t.message.Topic = name
+	return t
+}
+
+func (t *Client) MoodType(typeName string) *Client {
+	t.message.MoodType = typeName
+	return t
+}
+
+func (t *Client) Priority(priority float64) *Client {
+	if priority >= 1000 {
+		priority = 999
+	}
+	t.message.Priority = priority
+	return t
+}
+
+func (t *Client) Payload(payload []byte) *Client {
+	t.message.Payload = stringx.ByteToString(payload)
+	return t
+}
+
+func (t *Client) process(ctx context.Context, cmd IBaseCmd) error {
+
+	defer func() {
+		t.message = &Message{
```

Comment:
should not reset `t.message` here, it is not thread safe.

Reply:
done